virttest/utils_conn.py

```diff
@@ -1815,32 +1832,74 @@ def conn_recover(self):
         del self.client_saslconf
         del self.client_hosts
         del self.client_libvirtd_socket
-        # restart libvirtd service on server
-        client_session = self.client_session
         try:
-            libvirtd_service = utils_libvirtd.Libvirtd(session=client_session)
+            sessions = self.session_creator()
+            if not self.run_on_remote:
+                current_session = sessions[0]
+                runner = process.run
+            else:
+                current_session = sessions[1]
+                remote_runner = remote_old.RemoteRunner(session=current_session)
+                runner = remote_runner.run
+            libvirtd_service = utils_libvirtd.Libvirtd(session=current_session)
             if self.libvirt_ver:
                 # For traditional mode, need unmask libvirtd*.socket
                 if traditional_mode == 'yes':
-                    process.run("systemctl unmask libvirtd.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-admin.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-ro.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-tcp.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-tls.socket", shell=True)
+                    runner("systemctl unmask libvirtd.socket")
+                    runner("systemctl unmask libvirtd-admin.socket")
+                    runner("systemctl unmask libvirtd-ro.socket")
+                    runner("systemctl unmask libvirtd-tcp.socket")
+                    runner("systemctl unmask libvirtd-tls.socket")
                 else:
-                    process.run("systemctl daemon-reload", shell=True)
-                    process.run("systemctl stop libvirtd.socket", shell=True)
-                libvirtd_service.stop()
-                process.run("systemctl start libvirtd.socket", shell=True)
+                    runner("systemctl daemon-reload")
+                    runner("systemctl stop libvirtd.socket")
+                runner("systemctl stop libvirtd.service")
+                runner("systemctl start libvirtd.socket")
                 libvirtd_service.start()
             else:
                 libvirtd_service.restart()
+            current_session.close()
         except (remote.LoginError, aexpect.ShellError,
                 process.CmdError) as detail:
             raise ConnServerRestartError(detail)
 
         logging.debug("UNIX connection recover successfully.")
 
+    def session_creator(self):
+        """
+        Prepare a local/remote sessions as required and return them back.
+        """
+        server_ip = self.server_ip
+        client_ip = self.client_ip
+        server_user = self.server_user
+        client_user = self.client_user
+        server_pwd = self.server_pwd
+        client_pwd = self.client_pwd
+        if self.run_on_remote:
+            client_session = remote.wait_for_login('ssh',
```

Comment:
I think it looks like a duplicated sessio. Could you help to check if get_client_session() and get_server_session() meet your need?

Reply:
@dzhengfy It works on RHEL8.4, but I had issues on RHEL9.0, unfortunately cannot reproduce due to latest RHEL9.0 issues. What do you suggest? Should I keep try it on RHEL9.0 or update per working steps from 8.4?

Reply:
Updated as discussed, using class variables - self.client_session and self.server_session. It works on RHEL8.4 as shown below. Cannot verify on RHEL9.0 as there is known bug with SASL authentification at a moment.
Results:
<pre>avocado run --vt-type libvirt --vt-machine-type q35 virsh.remote_with_unix.positive_testing.allow_sasl_plain
WARNING:root:No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
WARNING:root:No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
<font color="#2A7BDE">JOB ID     : 2284f8e707c0c4f12f99e0ded16f9f59504bc517</font>
<font color="#2A7BDE">JOB LOG    : /root/avocado/job-results/job-2021-06-16T09.08-2284f8e/job.log</font>
 (1/1) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.allow_sasl_plain: <font color="#33DA7A">PASS</font> (70.33 s)
<font color="#2A7BDE">RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0</font>
<font color="#2A7BDE">JOB TIME   : 71.67 s</font>
</pre>